File: deploy/fastdeploy/cpu-gpu/python/det_keypoint_unite/det_keypoint_unite_infer.py
import fastdeploy as fd
import cv2
import os


def parse_arguments():
    import argparse
    import ast
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--tinypose_model_dir",
        required=True,
        help="path of paddletinypose model directory")
    parser.add_argument(
        "--det_model_dir", help="path of paddledetection model directory")
    parser.add_argument(
        "--image_file", required=True, help="path of test image file.")
    parser.add_argument(
        "--device",
        type=str,
        default='cpu',
        help="type of inference device, support 'cpu' or 'gpu'.")
    parser.add_argument(
        "--use_trt",
        type=ast.literal_eval,
        default=False,
        help="wether to use tensorrt.")
    return parser.parse_args()


def build_picodet_option(args):
    option = fd.RuntimeOption()

    if args.device.lower() == "gpu":
        option.use_gpu()

    if args.use_trt:
        option.use_paddle_infer_backend()
        # If use original Tensorrt, not Paddle-TensorRT,
        # please try `option.use_trt_backend()`
        option.paddle_infer_option.enable_trt = True
        option.paddle_infer_option.collect_trt_shape = True
        option.trt_option.set_shape("image", [1, 3, 320, 320], [1, 3, 320, 320],
                                             [1, 3, 320, 320])
        option.trt_option.set_shape("scale_factor", [1, 2], [1, 2], [1, 2])                                     
    return option


def build_tinypose_option(args):
    option = fd.RuntimeOption()

    if args.device.lower() == "gpu":
        option.use_gpu()

    if args.use_trt:
        option.use_paddle_infer_backend()
        # If use original Tensorrt, not Paddle-TensorRT,
        # please try `option.use_trt_backend()`
        option.paddle_infer_option.enable_trt = True
        option.paddle_infer_option.collect_trt_shape = True
        option.trt_option.set_shape("image", [1, 3, 256, 192], [1, 3, 256, 192],
                                             [1, 3, 256, 192])
    return option


args = parse_arguments()
picodet_model_file = os.path.join(args.det_model_dir, "model.pdmodel")
picodet_params_file = os.path.join(args.det_model_dir, "model.pdiparams")
picodet_config_file = os.path.join(args.det_model_dir, "infer_cfg.yml")

# setup runtime
runtime_option = build_picodet_option(args)
det_model = fd.vision.detection.PicoDet(
    picodet_model_file,
    picodet_params_file,
    picodet_config_file,
    runtime_option=runtime_option)

tinypose_model_file = os.path.join(args.tinypose_model_dir, "model.pdmodel")
tinypose_params_file = os.path.join(args.tinypose_model_dir, "model.pdiparams")
tinypose_config_file = os.path.join(args.tinypose_model_dir, "infer_cfg.yml")
# setup runtime
runtime_option = build_tinypose_option(args)
tinypose_model = fd.vision.keypointdetection.PPTinyPose(
    tinypose_model_file,
    tinypose_params_file,
    tinypose_config_file,
    runtime_option=runtime_option)

# predict
im = cv2.imread(args.image_file)
pipeline = fd.pipeline.PPTinyPose(det_model, tinypose_model)
pipeline.detection_model_score_threshold = 0.5
pipeline_result = pipeline.predict(im)
print("Paddle TinyPose Result:\n", pipeline_result)

# visualize
vis_im = fd.vision.vis_keypoint_detection(
    im, pipeline_result, conf_threshold=0.2)
cv2.imwrite("visualized_result.jpg", vis_im)
print("TinyPose visualized result save in ./visualized_result.jpg")

File: deploy/fastdeploy/cpu-gpu/cpp/det_keypoint_unite/det_keypoint_unite_infer.cc
// Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "fastdeploy/vision.h"
#include "fastdeploy/pipeline.h"

#ifdef WIN32
const char sep = '\\';
#else
const char sep = '/';
#endif

void CpuInfer(const std::string& det_model_dir,
              const std::string& tinypose_model_dir,
              const std::string& image_file) {
  auto det_model_file = det_model_dir + sep + "model.pdmodel";
  auto det_params_file = det_model_dir + sep + "model.pdiparams";
  auto det_config_file = det_model_dir + sep + "infer_cfg.yml";
  auto det_model = fastdeploy::vision::detection::PicoDet(
      det_model_file, det_params_file, det_config_file);
  if (!det_model.Initialized()) {
    std::cerr << "Detection Model Failed to initialize." << std::endl;
    return;
  }

  auto tinypose_model_file = tinypose_model_dir + sep + "model.pdmodel";
  auto tinypose_params_file = tinypose_model_dir + sep + "model.pdiparams";
  auto tinypose_config_file = tinypose_model_dir + sep + "infer_cfg.yml";
  auto tinypose_model = fastdeploy::vision::keypointdetection::PPTinyPose(
      tinypose_model_file, tinypose_params_file, tinypose_config_file);
  if (!tinypose_model.Initialized()) {
    std::cerr << "TinyPose Model Failed to initialize." << std::endl;
    return;
  }

  auto im = cv::imread(image_file);
  fastdeploy::vision::KeyPointDetectionResult res;

  auto pipeline =fastdeploy::pipeline::PPTinyPose(&det_model, &tinypose_model);
  pipeline.detection_model_score_threshold = 0.5;
  if (!pipeline.Predict(&im, &res)) {
    std::cerr << "TinyPose Prediction Failed." << std::endl;
    return;
  } else {
    std::cout << "TinyPose Prediction Done!" << std::endl;
  }
  
  std::cout << res.Str() << std::endl;

  auto vis_im =
      fastdeploy::vision::VisKeypointDetection(im, res, 0.2);
  cv::imwrite("vis_result.jpg", vis_im);
  std::cout << "TinyPose visualized result saved in ./vis_result.jpg"
            << std::endl;
}

void GpuInfer(const std::string& det_model_dir,
              const std::string& tinypose_model_dir,
              const std::string& image_file) {
  auto option = fastdeploy::RuntimeOption();
  option.UseGpu();
  auto det_model_file = det_model_dir + sep + "model.pdmodel";
  auto det_params_file = det_model_dir + sep + "model.pdiparams";
  auto det_config_file = det_model_dir + sep + "infer_cfg.yml";
  auto det_model = fastdeploy::vision::detection::PicoDet(
      det_model_file, det_params_file, det_config_file, option);
  if (!det_model.Initialized()) {
    std::cerr << "Detection Model Failed to initialize." << std::endl;
    return;
  }

  auto tinypose_model_file = tinypose_model_dir + sep + "model.pdmodel";
  auto tinypose_params_file = tinypose_model_dir + sep + "model.pdiparams";
  auto tinypose_config_file = tinypose_model_dir + sep + "infer_cfg.yml";
  auto tinypose_model = fastdeploy::vision::keypointdetection::PPTinyPose(
      tinypose_model_file, tinypose_params_file, tinypose_config_file, option);
  if (!tinypose_model.Initialized()) {
    std::cerr << "TinyPose Model Failed to initialize." << std::endl;
    return;
  }

  auto im = cv::imread(image_file);
  fastdeploy::vision::KeyPointDetectionResult res;

  auto pipeline =
      fastdeploy::pipeline::PPTinyPose(
          &det_model, &tinypose_model);
  pipeline.detection_model_score_threshold = 0.5;
  if (!pipeline.Predict(&im, &res)) {
    std::cerr << "TinyPose Prediction Failed." << std::endl;
    return;
  } else {
    std::cout << "TinyPose Prediction Done!" << std::endl;
  }
  
  std::cout << res.Str() << std::endl;

  auto vis_im =
      fastdeploy::vision::VisKeypointDetection(im, res, 0.2);
  cv::imwrite("vis_result.jpg", vis_im);
  std::cout << "TinyPose visualized result saved in ./vis_result.jpg"
            << std::endl;
}

void TrtInfer(const std::string& det_model_dir,
              const std::string& tinypose_model_dir,
              const std::string& image_file) {
  auto det_model_file = det_model_dir + sep + "model.pdmodel";
  auto det_params_file = det_model_dir + sep + "model.pdiparams";
  auto det_config_file = det_model_dir + sep + "infer_cfg.yml";

  auto det_option = fastdeploy::RuntimeOption();
  det_option.UseGpu();
  det_option.UsePaddleInferBackend();
  // If use original Tensorrt, not Paddle-TensorRT,
  // please try `option.UseTrtBackend()`
  det_option.paddle_infer_option.enable_trt = true;
  det_option.paddle_infer_option.collect_trt_shape = true;
  det_option.trt_option.SetShape("image", {1, 3, 320, 320}, {1, 3, 320, 320}, 
                                 {1, 3, 320, 320});
  det_option.trt_option.SetShape("scale_factor", {1, 2}, {1, 2}, {1, 2});    
  auto det_model = fastdeploy::vision::detection::PicoDet(
      det_model_file, det_params_file, det_config_file, det_option);
  if (!det_model.Initialized()) {
    std::cerr << "Detection Model Failed to initialize." << std::endl;
    return;
  }

  auto tinypose_model_file = tinypose_model_dir + sep + "model.pdmodel";
  auto tinypose_params_file = tinypose_model_dir + sep + "model.pdiparams";
  auto tinypose_config_file = tinypose_model_dir + sep + "infer_cfg.yml";
  auto tinypose_option = fastdeploy::RuntimeOption();
  
  tinypose_option.UseGpu();
  tinypose_option.UsePaddleInferBackend();
  // If use original Tensorrt, not Paddle-TensorRT,
  // please try `option.UseTrtBackend()`
  tinypose_option.paddle_infer_option.enable_trt = true;
  tinypose_option.paddle_infer_option.collect_trt_shape = true;
  tinypose_option.trt_option.SetShape("image", {1, 3, 256, 192}, {1, 3, 256, 192}, 
                                      {1, 3, 256, 192});
  auto tinypose_model = fastdeploy::vision::keypointdetection::PPTinyPose(
      tinypose_model_file, tinypose_params_file, tinypose_config_file,
      tinypose_option);
  if (!tinypose_model.Initialized()) {
    std::cerr << "TinyPose Model Failed to initialize." << std::endl;
    return;
  }

  auto im = cv::imread(image_file);
  fastdeploy::vision::KeyPointDetectionResult res;

  auto pipeline =
      fastdeploy::pipeline::PPTinyPose(
          &det_model, &tinypose_model);
  pipeline.detection_model_score_threshold = 0.5;
  if (!pipeline.Predict(&im, &res)) {
    std::cerr << "TinyPose Prediction Failed." << std::endl;
    return;
  } else {
    std::cout << "TinyPose Prediction Done!" << std::endl;
  }
  
  std::cout << res.Str() << std::endl;

  auto vis_im =
      fastdeploy::vision::VisKeypointDetection(im, res, 0.2);
  cv::imwrite("vis_result.jpg", vis_im);
  std::cout << "TinyPose visualized result saved in ./vis_result.jpg"
            << std::endl;
}

int main(int argc, char* argv[]) {
  if (argc < 5) {
    std::cout << "Usage: infer_demo path/to/detection_model_dir "
                 "path/to/pptinypose_model_dir path/to/image run_option, "
                 "e.g ./infer_model ./picodet_model_dir ./pptinypose_model_dir "
                 "./test.jpeg 0"
              << std::endl;
    std::cout << "The data type of run_option is int, 0: run with cpu; 1: run "
                 "with gpu; 2: run with gpu and use tensorrt backend;"
              << std::endl;
    return -1;
  }

  if (std::atoi(argv[4]) == 0) {
    CpuInfer(argv[1], argv[2], argv[3]);
  } else if (std::atoi(argv[4]) == 1) {
    GpuInfer(argv[1], argv[2], argv[3]);
  } else if (std::atoi(argv[4]) == 2) {
    TrtInfer(argv[1], argv[2], argv[3]);
  }
  return 0;
}
